### DISCUSSION TITLE
question: Does it support raster data?

### DISCUSSION THREAD

**Discussion by EverWinter23** [maplibre/martin#248] [Ideas]
Can we store/map raster data using `postgis + martin` to speed up rendering on the client side?

**Edit:** And if it does not, how can that be achieved?

**Comment by stepankuzmin** [comment]
Hi! Currently, martin doesn't support raster tiles, but it can be added in the feature releases.

**Comment by Baschdl** [comment]
@stepankuzmin Did you already think about how one could implement this? Resorting to mapnik etc. for rendering and just serving the resulting raster tiles?

**Reply by jferencik** [reply]
@Baschdl do you have any news on this feature?

@stepankuzmin, is the mapnik solution feasable in this case?

thanks

**Reply by Baschdl** [reply]
No, I didn't pursue this idea further

**Comment by EverWinter23** [comment]
@Baschdl Maybe this helps?

Take a look at `mbtiles` format. Use GDAL's `gdal2mbtiles` to tile the `GeoTiff` [raster source] and then use [mbtile-server](https://github.com/consbio/mbtileserver) to serve them to the frontend which can be consumed by `mapbox`, `OSM` etc.

**Comment by nyurik** [comment]
Update: Yes, Martin has been able to serve raster tiles from postgres, mbtiles, and pmtiles for a while now.

To try it out (if you have Rust installed, from the root of the Martin repo) - just do `cargo run tests/fixtures/files/` and there is a test file that contains jpgs -- so you can get a tile from `http://0.0.0.0:3000/geography-class-jpg/0/0/0`

If you have an mbtiles file with images, you can just run `martin myfile.mbtiles` and it should work too.

**Reply by ghost** [reply]
Hi there,
I can't find a resource, that explains how to serve raster data from a postgis raster source. Could you please elaborate.
Thanks Merlin

**Reply by AbelVM** [reply]
I'm interested in this topic too. I haven't found any mention in the docs about serving raster tiles, and it would fulfill all the requirements of my org. Is this achieved just [like this](https://github.com/maplibre/martin/blob/d85d95d5823407c2d8793df54b71a86c3b5e5c1c/tests/config.yaml#L160)? What about PostGIS rasters? There is no `geometry_column` to set in the yaml

Anyway, we should split this  discussion into 2 separate issues:
* Serving raster tiles that are actually "there" (stored in one of the supported data sources)
* Server-side rendering of vector info, and then serving the raster tiles (a la [TileServerGL](http://tileserver.org/))

**Reply by agcII** [reply]
any change?

**Reply by nyurik** [reply]
Once again -- this is already working.

I guess we could do some sort of a demo on how to server raster tiles, but at the end of the day, Martin does not care about the tile content, with the possible exception of how tilejson is constructed which already supports it.

To try it out (if you have Rust installed, from the root of the Martin repo) - just do `cargo run tests/fixtures/files/` and there is a test file that contains jpgs -- so you can get a tile from `http://0.0.0.0:3000/geography-class-jpg/0/0/0`

**Reply by sharkAndshark** [reply]
Maybe we could support raster in `PostGIS`? Like using the [ST_Tile](https://postgis.net/docs/en/RT_ST_Tile.html)?

**Reply by nyurik** [reply]
@sharkAndshark if someone wants to get raster from PostGIS, they can add a [pg function](https://maplibre.org/martin/sources-pg-functions.html) that Martin will auto-discover, and use a SQL comment that you implemented to indicate that the tile is a jpg. (comments are to be released soon).

I am not certain there is much value in supporting the auto-discover "serve from a table" approach like we do with MVTs. If you can come up with a good approach for auto-discovery for a raster-containing tables that would be generic enough, sure, lets discuss.

**Reply by sharkAndshark** [reply]
I guess some user have no time to learn or not familiar with writing `SQL function`. Before we konw the correct way, some thoughts :
* Give a guide/demo about using function with raster data in PostgreSQL
* Any plan to support [COG](https://www.cogeo.org/) or other formats?

**Reply by nyurik** [reply]
SQL functions are trivial to write for anyone already doing SQL,  i.e. they either know SQL (that's why they have a DB) -- so its not that difficult, or they don't know SQL, in which case it doesn't matter as they won't use a DB.  So in short - yes, please submit PRs for improving documentation on how to do this :smile: 

Sure, I am all for supporting formats like COG if anyone is interested in implementing it.  This is very similar to PMTiles support, and eventually we should position Martin as a web-native tile server (PRs welcome!)
